src/main/scala/symsim/examples/concrete/simplemaze/Maze.scala

```diff
@@ -0,0 +1,115 @@
+package symsim
+package examples.concrete.simplemaze
+
+import symsim.concrete.Randomized
+
+object Maze
+  extends Agent[MazeState, MazeFiniteState, MazeAction, MazeReward, Randomized] {
+
+    def isFinal (s: MazeState): Boolean =
+      s.x == 4 && (s.y == 2 || s.y == 3)
+
+    // Maze is already discrete 
+    def discretize (s: MazeState): MazeFiniteState =  { s }
+
+    private def mazeReward (s: MazeState) (a: MazeAction): MazeReward =
+      if (s == MazeState(x=4,y=3)) 10 //Good final state
+      else if (s == MazeState(x=4,y=2)) -10 // Bad final state
+      else -1
+
+    // Does this even make sense for the maze? 
+    /** Granularity of the step in seconds */
+    // private val t: Double = 1.0
+
+    // TODO: this is now deterministic but eventually needs to be randomized
+    def step (s: MazeState) (a: MazeAction): (MazeState, MazeReward) = { // Do cases over MazeAction
+      a match {
+        case Up =>    (stepUp(s),   1);
+        case Down =>  (stepDown(s), 1); 
+        case Left =>  (stepLeft(s), 1);
+        case Right => (stepRight(s),1);
+      }
+    }
+
+// I would like to select the step-function based on an enumeration type of actions
+    def stepUp(s: MazeState): MazeState = {
+      val y1 = if (s.x == 2 || s.y == 3)  (s.y) else s.y+1;
+      MazeState(x=s.x,y=y1)
+    }
+
+    def stepDown(s: MazeState): MazeState = {
+      val y1 = if (s.x == 2 || s.y == 1)  (s.y) else s.y-1
+      MazeState(x=s.x,y=y1)
+    }
+    
+    def stepLeft(s: MazeState): MazeState = {
+      val x1 = if (s.y == 2 || s.x == 1)  (s.x) else s.x-1
+      MazeState(x=x1,y=s.y)
+      }
+
+    def stepRight(s: MazeState): MazeState = {
+      val x1 = if (s == MazeState(y=1,x=2) || s.x == 4)  (s.x) else s.x-1
+      MazeState(x=x1,y=s.y)
+      }
+
+    def initialize: Randomized[MazeState] = for {
+      y <- Randomized.between (1, 3)
+      x <- Randomized.between (1, 4)
+      s0 = MazeState (x,y)
+      s <- if (isFinal (s0) || (s0 == MazeState(2,2))) initialize
+           else Randomized.const (s0)
+    } yield s
```

Comment:
def initialize: Randomized[MazeState] = for {
      y <- Randomized.between (1, 3)
      x <- Randomized.between (1, 4) if x != 2 && y != 2
      s0 = MazeState (x,y)
      s <- if (isFinal (s0)) initialize
           else Randomized.const (s0)
    } yield s

Reply:
Leaving this unchanged for now. When I tried the suggested solution, I got compilation errors such as 

value withFilter is not a member of symsim.concrete.Randomized[Int]

value withFilter is not a member of cats.data.IndexedStateT[cats.Eval,scala.util.Random,scala.util.Random,(Int, symsim.examples.concrete.simplemaze.MazeState)]